File: src/lib/pipeline.ts
// The multi-agent pipeline. Each agent uses one model for one focused task:
//
//   1. Fetcher       Gemini 2.5 Flash (grounded)   live news, multi-source
//   2. Classifier    Gemini 2.5 Flash Lite          domain + urgency + severity
//   3. Analyst       Gemini 2.5 Flash (grounded)    causes, connect-the-dots, narratives
//   4. Impact Mapper Gemini 2.5 Flash Lite          winners/losers + impact matrix
//   5. Historian     Groq Llama 4 Scout             history, blind spots, contrarian view
//   6. Synthesis     Groq Llama 4 Scout             debate: agreement vs tension
//   7. Rewriter      Groq Llama 4 Scout             final plain-English rewrite
//
// The exported orchestrators wire these agents together; the API routes and the
// scheduled scripts call the orchestrators.

import { runGemini, extractJson, MODEL_FLASH, MODEL_LITE } from "./gemini";
import { runGroq } from "./groq";
import * as P from "./prompts";
import type {
  AgentTake,
  DeepDive,
  QuickScan,
  RawStory,
  Severity,
  Source,
  Story,
  Synthesis,
} from "./types";

type StoryRef = Pick<Story, "headline" | "summary">;

function clampUrgency(n: unknown): number {
  const v = Math.round(Number(n));
  if (!Number.isFinite(v)) return 5;
  return Math.max(1, Math.min(10, v));
}

function normalizeSeverity(s: unknown): Severity {
  const v = String(s ?? "").toUpperCase();
  return v === "CRITICAL" || v === "HIGH" || v === "MEDIUM" || v === "LOW"
    ? v
    : "MEDIUM";
}

// --- AGENT 1: News Fetcher — Gemini 2.5 Flash + Google Search grounding ---
async function agentFetch(
  query?: string,
): Promise<{ raw: RawStory[]; sources: Source[] }> {
  const date = new Date().toISOString().slice(0, 10);
  const prompt = query?.trim()
    ? P.searchFetchPrompt(query.trim(), date)
    : P.newsFetchPrompt(date);

  const { text, sources } = await runGemini({
    system: P.FETCH_SYSTEM,
    prompt,
    model: MODEL_FLASH,
    grounding: true,
  });

  const parsed = extractJson<{ stories: RawStory[] }>(text);
  const raw: RawStory[] = (parsed.stories || []).slice(0, 5).map((s) => ({
    headline: String(s.headline ?? "Untitled"),
    summary: String(s.summary ?? ""),
    why: String(s.why ?? ""),
  }));
  if (raw.length === 0) {
    throw new Error("The fetcher agent returned no stories. Try again.");
  }
  return { raw, sources };
}

// --- AGENT 2: Classifier — Gemini 2.5 Flash Lite ---
async function agentClassify(
  raw: RawStory[],
): Promise<{ domain: string; urgency: number; severity: Severity }[]> {
  const { text } = await runGemini({
    system: P.CLASSIFY_SYSTEM,
    prompt: P.classifyPrompt(raw),
    model: MODEL_LITE,
  });
  const parsed = extractJson<{
    items: { domain: string; urgency: number; severity: string }[];
  }>(text);
  return (parsed.items || []).map((it) => ({
    domain: String(it.domain || "Society"),
    urgency: clampUrgency(it.urgency),
    severity: normalizeSeverity(it.severity),
  }));
}

// --- AGENT 3: Analyst — Gemini 2.5 Flash + grounding ---
async function agentAnalyze(story: StoryRef): Promise<{
  part: Pick<
    DeepDive,
    "rootCauses" | "butterflyChain" | "connectDots" | "narratives"
  >;
  sources: Source[];
}> {
  const { text, sources } = await runGemini({
    system: P.ANALYZE_SYSTEM,
    prompt: P.analyzePrompt(story),
    model: MODEL_FLASH,
    grounding: true,
  });
  const part = extractJson<
    Pick<DeepDive, "rootCauses" | "butterflyChain" | "connectDots" | "narratives">
  >(text);
  return { part, sources };
}

// --- AGENT 4: Impact Mapper — Gemini 2.5 Flash Lite ---
async function agentImpact(
  story: StoryRef,
): Promise<Pick<DeepDive, "impacts" | "winners" | "losers">> {
  const { text } = await runGemini({
    system: P.ANALYZE_SYSTEM,
    prompt: P.impactPrompt(story),
    model: MODEL_LITE,
  });
  return extractJson<Pick<DeepDive, "impacts" | "winners" | "losers">>(text);
}

// --- AGENT 5: Historian + contrarian — Groq Llama 4 Scout ---
async function agentHistory(
  story: StoryRef,
): Promise<
  Pick<DeepDive, "historicalPattern" | "blindSpots" | "predictions" | "contrarianView">
> {
  const text = await runGroq({
    system: P.HISTORY_SYSTEM,
    prompt: P.historyPrompt(story),
  });
  return extractJson<
    Pick<
      DeepDive,
      "historicalPattern" | "blindSpots" | "predictions" | "contrarianView"
    >
  >(text);
}

// --- Multi-Agent View: 5 analyst perspectives — Gemini 2.5 Flash + grounding ---
async function agentPerspectives(
  story: StoryRef,
): Promise<{ agents: AgentTake[]; sources: Source[] }> {
  const { text, sources } = await runGemini({
    system: P.PERSPECTIVES_SYSTEM,
    prompt: P.perspectivesPrompt(story),
    model: MODEL_FLASH,
    grounding: true,
  });
  const parsed = extractJson<{ agents: AgentTake[] }>(text);
  return { agents: parsed.agents || [], sources };
}

// --- AGENT 6: Debate synthesis — Groq Llama 4 Scout ---
async function agentSynthesis(agents: AgentTake[]): Promise<Synthesis> {
  const text = await runGroq({
    system: P.SYNTHESIS_SYSTEM,
    prompt: P.synthesisPrompt(JSON.stringify({ agents })),
  });
  return extractJson<Synthesis>(text);
}

// --- AGENT 7: Rewriter — Groq Llama 4 Scout ---
// Rewrites every text value into plain English while keeping the JSON shape.
// If the rewrite drifts or fails, the original data is returned unchanged.
async function agentRewrite<T>(label: string, data: T): Promise<T> {
  try {
    const text = await runGroq({
      system: P.REWRITE_SYSTEM,
      prompt: P.rewritePrompt(label, JSON.stringify(data)),
    });
    return extractJson<T>(text);
  } catch {
    return data;
  }
}

// ===================== ORCHESTRATORS =====================

/**
 * Today's important news, or the top angles on a topic when `query` is given.
 * Fetcher → (Classifier ‖ Rewriter) → merged, ranked, severity-tagged stories.
 */
export async function fetchStories(
  query?: string,
): Promise<{ stories: Story[]; sources: Source[] }> {
  const { raw, sources } = await agentFetch(query);

  const [classes, rewritten] = await Promise.all([
    agentClassify(raw),
    agentRewrite<{ stories: RawStory[] }>("set of news stories", {
      stories: raw,
    }),
  ]);

  const stamp = Date.now();
  const stories: Story[] = raw.map((r, i) => {
    const rw = rewritten.stories?.[i];
    const cls = classes[i];
    return {
      id: `s${stamp}-${i}`,
      headline: rw?.headline || r.headline,
      summary: rw?.summary || r.summary,
      why: rw?.why || r.why,
      domain: cls?.domain || "Society",
      urgency: cls?.urgency ?? 5,
      severity: cls?.severity ?? "MEDIUM",
    };
  });
  stories.sort((a, b) => b.urgency - a.urgency);

  return { stories, sources };
}

/**
 * Full deep-dive analysis of one story.
 * (Analyst ‖ Impact Mapper ‖ Historian) → merged → Rewriter.
 */
export async function buildDeepDive(
  story: StoryRef,
): Promise<{ analysis: DeepDive; sources: Source[] }> {
  const [analysis, impact, history] = await Promise.all([
    agentAnalyze(story),
    agentImpact(story),
    agentHistory(story),
  ]);

  const merged: DeepDive = {
    rootCauses: analysis.part.rootCauses,
    butterflyChain: analysis.part.butterflyChain || [],
    connectDots: analysis.part.connectDots,
    narratives: analysis.part.narratives || [],
    impacts: impact.impacts || [],
    winners: impact.winners || [],
    losers: impact.losers || [],
    historicalPattern: history.historicalPattern,
    blindSpots: history.blindSpots || [],
    predictions: history.predictions || [],
    contrarianView: history.contrarianView,
  };

  const final = await agentRewrite<DeepDive>("deep-dive analysis", merged);
  return { analysis: final, sources: analysis.sources };
}

/**
 * Five specialist analysts weigh in, then a synthesis agent finds the debate.
 * Perspectives → Synthesis → Rewriter.
 */
export async function buildAgentTakes(
  story: StoryRef,
): Promise<{ agents: AgentTake[]; synthesis: Synthesis; sources: Source[] }> {
  const { agents, sources } = await agentPerspectives(story);
  const synthesis = await agentSynthesis(agents);
  const final = await agentRewrite<{ agents: AgentTake[]; synthesis: Synthesis }>(
    "panel of analyst views and their debate",
    { agents, synthesis },
  );
  return {
    agents: final.agents || agents,
    synthesis: final.synthesis || synthesis,
    sources,
  };
}

/**
 * Lightweight breaking-news scan for the every-2-hours monitor — one grounded
 * Gemini call. Returns whether anything CRITICAL is breaking right now.
 */
export async function quickScan(): Promise<QuickScan> {
  const date = new Date().toISOString().slice(0, 10);
  const { text } = await runGemini({
    system: P.MONITOR_SYSTEM,
    prompt: P.monitorPrompt(date),
    model: MODEL_FLASH,
    grounding: true,
  });
  const parsed = extractJson<QuickScan>(text);
  return {
    critical: Boolean(parsed.critical),
    items: (parsed.items || [])
      .slice(0, 5)
      .map((it) => ({
        headline: String(it.headline ?? ""),
        summary: String(it.summary ?? ""),
        why: String(it.why ?? ""),
      }))
      .filter((it) => it.headline),
  };
}

File: app/pipeline.py
"""
The multi-agent pipeline. Each agent uses one model for one focused task:

  1. Fetcher       Gemini 2.5 Flash (grounded)   live news, multi-source
  2. Classifier    Gemini 2.5 Flash Lite          domain + urgency + severity
  3. Analyst       Gemini 2.5 Flash (grounded)    causes, connect-the-dots, narratives
  4. Impact Mapper Gemini 2.5 Flash Lite          winners/losers + impact matrix
  5. Historian     Groq Llama 4 Scout             history, blind spots, contrarian view
  6. Synthesis     Groq Llama 4 Scout             debate: agreement vs tension
  7. Rewriter      Groq Llama 4 Scout             final plain-English rewrite

The exported orchestrators wire these agents together; the API routes and the
scheduled scripts call the orchestrators. Independent agents run in parallel.
"""

from __future__ import annotations

import time
from concurrent.futures import ThreadPoolExecutor
from datetime import date

from app import prompts as P
from app.gemini import MODEL_FLASH, MODEL_LITE, extract_json, run_gemini
from app.groq_client import run_groq

SEVERITIES = {"CRITICAL", "HIGH", "MEDIUM", "LOW"}


def _clamp_urgency(value) -> int:
    try:
        v = round(float(value))
    except (TypeError, ValueError):
        return 5
    return max(1, min(10, v))


def _normalize_severity(value) -> str:
    v = str(value or "").upper()
    return v if v in SEVERITIES else "MEDIUM"


# --- AGENT 1: News Fetcher — Gemini 2.5 Flash + Google Search grounding ---
def _agent_fetch(query: str | None) -> dict:
    today = date.today().isoformat()
    prompt = (
        P.search_fetch_prompt(query.strip(), today)
        if query and query.strip()
        else P.news_fetch_prompt(today)
    )
    result = run_gemini(P.FETCH_SYSTEM, prompt, model=MODEL_FLASH, grounding=True)
    parsed = extract_json(result["text"])
    raw = [
        {
            "headline": str(s.get("headline", "Untitled")),
            "summary": str(s.get("summary", "")),
            "why": str(s.get("why", "")),
        }
        for s in (parsed.get("stories") or [])[:5]
    ]
    if not raw:
        raise RuntimeError("The fetcher agent returned no stories. Try again.")
    return {"raw": raw, "sources": result["sources"]}


# --- AGENT 2: Classifier — Gemini 2.5 Flash Lite ---
def _agent_classify(raw: list[dict]) -> list[dict]:
    result = run_gemini(P.CLASSIFY_SYSTEM, P.classify_prompt(raw), model=MODEL_LITE)
    parsed = extract_json(result["text"])
    return [
        {
            "domain": str(it.get("domain") or "Society"),
            "urgency": _clamp_urgency(it.get("urgency")),
            "severity": _normalize_severity(it.get("severity")),
        }
        for it in (parsed.get("items") or [])
    ]


# --- AGENT 3: Analyst — Gemini 2.5 Flash + grounding ---
def _agent_analyze(story: dict) -> dict:
    result = run_gemini(
        P.ANALYZE_SYSTEM, P.analyze_prompt(story), model=MODEL_FLASH, grounding=True
    )
    return {"part": extract_json(result["text"]), "sources": result["sources"]}


# --- AGENT 4: Impact Mapper — Gemini 2.5 Flash Lite ---
def _agent_impact(story: dict) -> dict:
    result = run_gemini(P.ANALYZE_SYSTEM, P.impact_prompt(story), model=MODEL_LITE)
    return extract_json(result["text"])


# --- AGENT 5: Historian + contrarian — Groq Llama 4 Scout ---
def _agent_history(story: dict) -> dict:
    return extract_json(run_groq(P.HISTORY_SYSTEM, P.history_prompt(story)))


# --- Multi-Agent View: 5 analyst perspectives — Gemini 2.5 Flash + grounding ---
def _agent_perspectives(story: dict) -> dict:
    result = run_gemini(
        P.PERSPECTIVES_SYSTEM,
        P.perspectives_prompt(story),
        model=MODEL_FLASH,
        grounding=True,
    )
    parsed = extract_json(result["text"])
    return {"agents": parsed.get("agents") or [], "sources": result["sources"]}


# --- AGENT 6: Debate synthesis — Groq Llama 4 Scout ---
def _agent_synthesis(agents: list[dict]) -> dict:
    return extract_json(run_groq(P.SYNTHESIS_SYSTEM, P.synthesis_prompt({"agents": agents})))


# --- AGENT 7: Rewriter — Groq Llama 4 Scout ---
def _agent_rewrite(label: str, data):
    """Rewrite text values into plain English; return the original on failure."""
    import json

    try:
        return extract_json(run_groq(P.REWRITE_SYSTEM, P.rewrite_prompt(label, json.dumps(data))))
    except Exception:
        return data


# ===================== ORCHESTRATORS =====================


def fetch_stories(query: str | None = None) -> dict:
    """Today's important news, or the top angles on a topic when `query` is given."""
    fetched = _agent_fetch(query)
    raw = fetched["raw"]

    with ThreadPoolExecutor(max_workers=2) as pool:
        classes_future = pool.submit(_agent_classify, raw)
        rewrite_future = pool.submit(
            _agent_rewrite, "set of news stories", {"stories": raw}
        )
        classes = classes_future.result()
        rewritten = rewrite_future.result()

    stamp = int(time.time() * 1000)
    rewritten_stories = rewritten.get("stories") or []
    stories: list[dict] = []
    for i, r in enumerate(raw):
        rw = rewritten_stories[i] if i < len(rewritten_stories) else {}
        cls = classes[i] if i < len(classes) else {}
        stories.append(
            {
                "id": f"s{stamp}-{i}",
                "headline": rw.get("headline") or r["headline"],
                "summary": rw.get("summary") or r["summary"],
                "why": rw.get("why") or r["why"],
                "domain": cls.get("domain") or "Society",
                "urgency": cls.get("urgency", 5),
                "severity": cls.get("severity") or "MEDIUM",
            }
        )

    stories.sort(key=lambda s: s["urgency"], reverse=True)
    return {"stories": stories, "sources": fetched["sources"]}


def build_deep_dive(story: dict) -> dict:
    """Full deep-dive analysis of one story. (Analyst ‖ Impact ‖ Historian) → Rewriter."""
    with ThreadPoolExecutor(max_workers=3) as pool:
        analyze_future = pool.submit(_agent_analyze, story)
        impact_future = pool.submit(_agent_impact, story)
        history_future = pool.submit(_agent_history, story)
        analysis = analyze_future.result()
        impact = impact_future.result()
        history = history_future.result()

    part = analysis["part"]
    merged = {
        "rootCauses": part.get("rootCauses"),
        "butterflyChain": part.get("butterflyChain") or [],
        "connectDots": part.get("connectDots"),
        "narratives": part.get("narratives") or [],
        "impacts": impact.get("impacts") or [],
        "winners": impact.get("winners") or [],
        "losers": impact.get("losers") or [],
        "historicalPattern": history.get("historicalPattern"),
        "blindSpots": history.get("blindSpots") or [],
        "predictions": history.get("predictions") or [],
        "contrarianView": history.get("contrarianView"),
    }
    final = _agent_rewrite("deep-dive analysis", merged)
    return {"analysis": final, "sources": analysis["sources"]}


def build_agent_takes(story: dict) -> dict:
    """Five specialist analysts weigh in, then a synthesis agent finds the debate."""
    perspectives = _agent_perspectives(story)
    agents = perspectives["agents"]
    synthesis = _agent_synthesis(agents)
    final = _agent_rewrite(
        "panel of analyst views and their debate",
        {"agents": agents, "synthesis": synthesis},
    )
    return {
        "agents": final.get("agents") or agents,
        "synthesis": final.get("synthesis") or synthesis,
        "sources": perspectives["sources"],
    }


def build_verdict(topic: str, story: dict) -> dict:
    """Deep Signal's own opinion on a topic — analysis, stance, solution, outcomes."""
    v = extract_json(run_groq(P.VERDICT_SYSTEM, P.verdict_prompt(topic, story)))
    outcomes = [
        {"horizon": str(o.get("horizon", "")), "outcome": str(o.get("outcome", ""))}
        for o in (v.get("outcomes") or [])[:4]
        if o.get("outcome")
    ]
    return {
        "analysis": str(v.get("analysis", "")),
        "opinion": str(v.get("opinion", "")),
        "solution": str(v.get("solution", "")),
        "outcomes": outcomes,
    }


def quick_scan() -> dict:
    """Lightweight breaking-news scan for the monitor — one grounded Gemini call."""
    today = date.today().isoformat()
    result = run_gemini(
        P.MONITOR_SYSTEM, P.monitor_prompt(today), model=MODEL_FLASH, grounding=True
    )
    parsed = extract_json(result["text"])
    items = [
        {
            "headline": str(it.get("headline", "")),
            "summary": str(it.get("summary", "")),
            "why": str(it.get("why", "")),
        }
        for it in (parsed.get("items") or [])[:5]
        if it.get("headline")
    ]
    return {"critical": bool(parsed.get("critical")), "items": items}
